File: src/ffi.erl
-module(ffi).

-export([
    parse_adv/1
]).

% Parses mirego/elixir-security-advisories advisory yaml
% files and returns these fields in an array
% - id: binary()
% - package: binary()
% - severity: binary()
% - title: binary()
% - vulnerable_version_ranges: list(binary())
%
% @param {binary()} content
% @returns {ok, {binary(), binary(), binary(), binary(), list(binary())}}
parse_adv(Raw) ->
    try do_parse_adv(Raw) of
        Res -> Res
    catch
        _:_ -> {error, nil}
    end.

do_parse_adv(Raw) ->
    maybe
        [Content] ?= yamerl:decode(Raw),
        {_, ID} ?= lists:keyfind("id", 1, Content),
        {_, Name} ?= lists:keyfind("package", 1, Content),
        {_, Severity} ?= lists:keyfind("severity", 1, Content),
        {_, Title} ?= lists:keyfind("title", 1, Content),
        {_, Vulnerable_version_ranges} ?= lists:keyfind("vulnerable_version_ranges", 1, Content),
        ID_parsed ?= unicode:characters_to_binary(ID),
        Name_parsed ?= unicode:characters_to_binary(Name),
        Severity_parsed ?= unicode:characters_to_binary(Severity),
        Title_parsed ?= unicode:characters_to_binary(Title),
        Vulnerable_version_ranges_parsed ?= lists:map(
            fun unicode:characters_to_binary/1, Vulnerable_version_ranges
        ),
        true ?= is_binary(ID_parsed),
        true ?= is_binary(Name_parsed),
        true ?= is_binary(Severity_parsed),
        true ?= is_binary(Title_parsed),
        true ?= lists:all(fun is_binary/1, Vulnerable_version_ranges_parsed),
        {ok, {
            ID_parsed,
            Name_parsed,
            Severity_parsed,
            Title_parsed,
            Vulnerable_version_ranges_parsed
        }}
    else
        _ -> {error, nil}
    end.


    

File: src/ffi.mjs
import { parse as parse_yaml } from "yaml";
import { Error, Ok, toList } from "../prelude.mjs";

const string_type = "string";
const err = new Error(null);

/**
 * Parses mirego/elixir-security-advisories advisory yaml
 * files and returns these fields in an array
 * - id: string
 * - package: string
 * - severity: string
 * - title: string
 * - vulnerable_version_ranges: List<string>
 *
 * @param {string} content
 * @returns Result<List<String | List<String>>, null>
 */
export function parse_adv(content) {
  try {
    const {
      id,
      package: pkg,
      severity,
      title,
      vulnerable_version_ranges,
    } = parse_yaml(content);

    if (!id || !pkg || !severity || !title || !vulnerable_version_ranges) {
      return err;
    }

    if (typeof id !== string_type) {
      return err;
    }

    if (typeof pkg !== string_type) {
      return err;
    }

    if (typeof severity !== string_type) {
      return err;
    }

    if (typeof title !== string_type) {
      return err;
    }

    if (!Array.isArray(vulnerable_version_ranges)) {
      return err;
    }

    for (const element of vulnerable_version_ranges) {
      if (typeof element !== string_type) {
        return err;
      }
    }

    return new Ok([
      id,
      pkg,
      severity,
      title,
      toList(vulnerable_version_ranges),
    ]);
  } catch {
    return err;
  }
}
